Stop @EnableWebMvc from being automatically added to Grails Applications

codeconsole opened the pull request:
Fixes #13843

This produces unknown/unexpected behavior and should not be done.  @EnableWebMvc is a Spring Framework configuration annotation not to be used with Spring Boot.  It produces undesired behavior by automatically creating beans that are not wanted. WebMvcAutoConfiguration performs all the behavior of this configuration because it extends it and adds conditional logic. 

Fixes 
```
Overriding bean definition for bean 'localeResolver' with a different definition: replacing [Root bean: class=null; scope=; abstract=false; lazyInit=null; autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; fallback=false; factoryBeanName=org.springframework.web.servlet.config.annotation.DelegatingWebMvcConfiguration; factoryMethodName=localeResolver; initMethodNames=null; destroyMethodNames=[(inferred)]; defined in org.springframework.web.servlet.config.annotation.DelegatingWebMvcConfiguration] with [Generic bean: class=org.springframework.web.servlet.i18n.SessionLocaleResolver; scope=singleton; abstract=false; lazyInit=false; autowireMode=0; dependencyCheck=0; autowireCandidate=true; primary=false; fallback=false; factoryBeanName=null; factoryMethodName=null; initMethodNames=null; destroyMethodNames=null]
Overriding bean definition for bean 'org.springframework.aop.config.internalAutoProxyCreator' with a different definition: replacing [Root bean: class=org.springframework.aop.framework.autoproxy.InfrastructureAdvisorAutoProxyCreator; scope=; abstract=false; lazyInit=null; autowireMode=0; dependencyCheck=0; autowireCandidate=true; primary=false; fallback=false; factoryBeanName=null; factoryMethodName=null; initMethodNames=null; destroyMethodNames=null] with [Generic bean: class=org.grails.spring.aop.autoproxy.GroovyAwareInfrastructureAdvisorAutoProxyCreator; scope=singleton; abstract=false; lazyInit=false; autowireMode=0; dependencyCheck=0; autowireCandidate=true; primary=false; fallback=false; factoryBeanName=null; factoryMethodName=null; initMethodNames=null; destroyMethodNames=null]
```

```java
class WebMvcAutoConfiguration {
    static class EnableWebMvcConfiguration extends DelegatingWebMvcConfiguration { 
    } 
}

class DelegatingWebMvcConfiguration extends WebMvcConfigurationSupport {
}

class WebMvcConfigurationSupport extends WebMvcConfigurationSupport {
    @Bean public LocaleResolver localeResolver() { }
}

```

vs

```java
@Import(DelegatingWebMvcConfiguration.class)
public @interface EnableWebMvc { }
```

---

With `@EnableWebMvc` gone, Boot's `WebMvcAutoConfiguration` is now active for Grails servlet web apps. This PR adapts Grails so the change is transparent:

1. **Remove the auto-injected `@EnableWebMvc`** (`ApplicationClassInjector`).
2. **Default view resolver** — Boot now contributes a catch-all `defaultViewResolver` that breaks REST/JSON-views apps with `Circular view path` errors. New `GrailsViewResolverAutoConfiguration` removes it centrally for every Grails servlet web app, controlled by `grails.web.removeDefaultViewResolverBean` (default `true`). This replaces the GSP-only `spring.gsp.removeDefaultViewResolverBean`; `spring.gsp.replaceViewResolverBean` is unchanged.
3. **Request binding** — `GrailsWebRequestFilter` now extends `RequestContextFilter` and is exposed as a `RequestContextFilter` bean so Boot's `WebMvcAutoConfiguration` backs off its own `OrderedRequestContextFilter`, which would otherwise rebind a plain `ServletRequestAttributes` between Grails' filter and the Spring Security chain and cause a `GrailsWebRequest` `ClassCastException`.

**Docs** — upgrade-notes section "27. @EnableWebMvc No Longer Added to Grails Applications" in `upgrading80x.adoc`.

**Tests** — `ControllersAutoConfigurationSpec` verifies the filter is exposed as a `RequestContextFilter` at the Grails request-filter order, and that Boot's `requestContextFilter` actually backs off (with a contrast case proving it's present without the Grails auto-config).

# Backward Compatibility

For legacy apps dependent on previous configuration, restoration is possible by simply adding `@EnableWebMvc` manually to `Application.groovy` 
```groovy
import org.springframework.web.servlet.config.annotation.EnableWebMvc;

@EnableWebMvc
class Application extends GrailsAutoConfiguration {
// ...
```

codeconsole committed:
Stop @EnableWebMvc from being automatically added to Grails Applications

codeconsole commented:
https://github.com/grails/grails-core/issues/13843

matrei commented:
Did you try this out with `grails-functional-tests`?
I get the following errors:

```console
BookFunctionalSpec > Test that when the /viewBooks URL is hit it redirects to the book list FAILED
    Condition not satisfied:

    title == "Book List"
    |     |
    |     false
    |     36 differences (7% similarity)
    |     (HTTP) (Statu)s( 500 – In)t(ernal Server Error)
    |     (Book) (Li---)s(---------)t(------------------)
    HTTP Status 500 – Internal Server Error
        at functionaltests.BookFunctionalSpec.Test that when the /viewBooks URL is hit it redirects to the book list(BookFunctionalSpec.groovy:15)
```

```console
'2024-11-20 17:31:50.853 ERROR --- [nio-8080-exec-1] .C.[.[.[.[dispatcherServletRegistration] : Servlet.service() for servlet [dispatcherServletRegistration] in context with path [] threw exception

org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object 'org.sitemesh.webapp.contentfilter.HttpServletRequestFilterable@6b88ebe0' with class 'org.springframework.web.context.request.ServletRequestAttributes' to class 'org.grails.web.servlet.mvc.GrailsWebRequest'
	at org.codehaus.groovy.runtime.typehandling.DefaultTypeTransformation.continueCastOnSAM(DefaultTypeTransformation.java:415)
```

codeconsole commented:
> Did you try this out with `grails-functional-tests`? I get the following errors:

@matrei please read above "This pull request is not finished"

matrei commented:
Sorry, I thought I got a review request.

codeconsole commented:
> Sorry, I thought I got a review request.

@matrei well, technically you did, but I immediately cancelled it and added the line "this pull request is not finished" before you got here😜

jeffscottbrown commented:
> This pull request is not finished

I have converted this PR to Draft for now.

jdaugherty commented:
This is now been here for almost a year and half with no progress.  I'm going to close the PR for now - please reopen when you feel it's ready for review.

sbglasius reviewed (commented):
I am not sure, I understand the full impact of this PR; but the code looks sound and if tests passes, I'm good.

I would however, like to see it documented in the upgrade notes, so that users can figure out, why their application might fail, if that would ever happen?

codeconsole commented:
@sbglasius thanks for the review — I've documented this so users can see why an app might be affected and how to control it:

- **Upgrade notes**: new section *"27. @EnableWebMvc No Longer Added to Grails Applications"* in `grails-doc` (`upgrading80x.adoc`). It covers the two user-visible behaviors and their opt-outs — the `Circular view path` symptom from Boot's `defaultViewResolver` (`grails.web.removeDefaultViewResolverBean`, default `true`) and the `GrailsWebRequest` binding.
- **PR description**: updated to spell out the full scope (the `defaultViewResolver` consolidation and the `GrailsWebRequestFilter` → `RequestContextFilter` change), since the original only mentioned the `@EnableWebMvc` removal — which I think is what made the impact hard to gauge.
- **Tests**: added `ControllersAutoConfigurationSpec`, which proves Boot's `requestContextFilter` actually backs off (with a contrast case showing it's present without the Grails auto-config), so the binding behavior is regression-guarded.

jdaugherty commented on grails-controllers/build.gradle:
```diff
@@ -63,6 +63,10 @@ dependencies {
 
     testImplementation 'jakarta.servlet:jakarta.servlet-api'
     testImplementation 'org.springframework:spring-test'
+    // WebApplicationContextRunner verifies Boot's WebMvcAutoConfiguration backs off its RequestContextFilter
```
Remove the comments around dependenices

codeconsole replied:
Removed in 12505aa.

codeconsole replied:
Removed — dropped the empty configuration closure and its comment on the `spring-test` `compileOnly` dependency.

jdaugherty commented on grails-controllers/src/main/groovy/org/grails/plugins/web/controllers/ControllersAutoConfiguration.java:
```diff
@@ -108,13 +108,24 @@ public FilterRegistrationBean<Filter> hiddenHttpMethodFilter() {
         return registrationBean;
     }
 
+    // Exposed as a RequestContextFilter bean (GrailsWebRequestFilter extends it) so that Boot's
```
Simplify the comment

codeconsole replied:
Simplified in 12505aa.

codeconsole replied:
Simplified — condensed to the two essential points (`GrailsWebRequestFilter` is a `RequestContextFilter` so Boot backs off, and the registration-bean-name gate so a user overriding only that registration makes this raw filter back off with it).

jdaugherty reviewed (commented):
We need to merge the security project prior to merging this.  From what I can tell, it relies on WebMvcAutoConfiguration and this is going to break that code.

jdaugherty commented:
Merged the latest 8.x changes, which should resolve failing tests.

jdaugherty commented:
I had Fabled take a look at this PR now that we've merged the security changes.  I think the issue identified in the security PR is still a problem.  We're now registering beans that we don't want registered because using the MVC annotation effectively prevented spring boot from registering from defaults that Grails relied on being absent.  Here are the details: 

**Impacts beyond what the PR/docs cover**

This is the important part: activating WebMvcAutoConfiguration brings in more than the two things the PR compensates for. These are behavior changes every Grails 8 app will see:

- OrderedFormContentFilter is now registered (on by default via spring.mvc.formcontent.filter.enabled). PUT/PATCH/DELETE form-encoded bodies become visible as request parameters — previously they weren't. Arguably an improvement, but it's a real behavioral difference for apps with PUT/PATCH endpoints doing their own body parsing, and it's not mentioned in the upgrade notes.
- spring.mvc.* and spring.web.* properties now take effect. Any app that had these set (inertly, copied from Boot docs) will see them suddenly activate on upgrade: spring.web.locale, spring.mvc.format.date, spring.web.resources.*, etc.
- Boot's static-resource handling and welcome-page mapping are now live. Boot's WebMvcAutoConfigurationAdapter registers a /** resource handler (classpath:/static, /public, …) alongside Grails' own GrailsWebMvcConfigurer handler, plus WelcomePageHandlerMapping. URLs that previously 404'd through Grails URL-mapping error handling may now be served as static resources or an index page — this can change error-page behavior and even shadow expectations in apps with catch-all URL mappings.
- viewResolver is now Boot's ContentNegotiatingViewResolver for non-GSP apps (GSP apps alias it to gspViewResolver as before). Generally compatible since it delegates, but it's a different resolution pipeline than before.
- mvcConversionService is now Boot's ApplicationConversionService (honoring format properties) instead of Spring's plain one, and message converters get the full Boot customization (context ObjectMapper, etc.) — subtle differences for @RestController-style endpoints and MVC-layer binding.

I'd suggest a short bullet list of these in section 31 of the upgrade notes ("other Boot MVC features that become active"), since the current text says "No action is required for most applications" and only documents the two compensated differences.

**One real edge-case bug candidate**

If a user overrides only the grailsWebRequestFilter registration bean (which the docs explicitly bless), the auto-configured grailsWebRequest plain Filter bean is still created (@ConditionalOnMissingBean(GrailsWebRequestFilter.class) doesn't back off), and since no registration bean references it anymore, Boot's adaptable-beans mechanism will auto-register it on /* at Ordered.LOWEST_PRECEDENCE — running in addition to the user's replacement filter. Previously, overriding the registration bean fully removed the Grails filter. The new spec's override test only asserts bean identity in the context, so it wouldn't catch this. Fix options: gate grailsWebRequest on the registration bean name too, or have the override test assert against the actual servlet filter chain.

codeconsole committed:
Merge branch '8.0.x' into 7.0.x-remove-EnableWebMvc

codeconsole committed:
Merge branch '8.0.x' into 7.0.x-remove-EnableWebMvc

codeconsole committed:
Prevent grailsWebRequest double-registration; document Boot MVC side effects

Addresses review feedback on the @EnableWebMvc removal.

- Gate the grailsWebRequest raw filter bean on the "grailsWebRequestFilter"
  registration bean name as well as its type. Overriding only the registration
  bean now makes the raw filter back off with it, instead of leaving it
  unwrapped for Boot's adaptable-beans machinery to auto-register on "/*" at
  Ordered.LOWEST_PRECEDENCE — a second, mis-ordered copy of the filter.

- Strengthen ControllersAutoConfigurationSpec to assert the assembled servlet
  filter chain (via ServletContextInitializerBeans): the override case now
  verifies no stray GrailsWebRequestFilter is registered, and the default case
  verifies exactly one is. These fail without the gating fix.

- Add upgrade-notes section 31.3 documenting the Boot WebMvcAutoConfiguration
  features that become active with @EnableWebMvc gone: OrderedFormContentFilter,
  spring.mvc.* / spring.web.* properties, static-resource handling and welcome
  page, ContentNegotiatingViewResolver, and ApplicationConversionService.

codeconsole commented:
Thanks @jdaugherty (and Fabled) — both are spot on, and I've updated this PR to address them.

Quick context on the diff: the original `@EnableWebMvc` removal already landed on `8.0.x` when #15916 merged this branch, so the diff GitHub renders here (computed against the old merge base) overstates what's actually new. I've merged current `8.0.x` back in and added the two follow-ups on top, so the real delta this PR now carries is just those fixes.

**1. Undocumented `WebMvcAutoConfiguration` side effects — documented.** Added a new "31.3 Other Boot MVC features now active" subsection to the upgrade notes covering:
- `OrderedFormContentFilter` (default-on via `spring.mvc.formcontent.filter.enabled`) — PUT/PATCH/DELETE form-encoded bodies now surface as request parameters.
- `spring.mvc.*` / `spring.web.*` properties now taking effect (`spring.web.locale`, `spring.mvc.format.*`, `spring.web.resources.*`, …).
- Boot's `/**` static-resource handler + `WelcomePageHandlerMapping`, and their effect on 404 / catch-all URL-mapping behavior (with the `spring.web.resources.add-mappings` opt-out).
- `viewResolver` now `ContentNegotiatingViewResolver` for non-GSP apps (GSP unchanged).
- `mvcConversionService` now `ApplicationConversionService`, plus Boot's message-converter customization.

**2. `grailsWebRequest` double-registration — fixed.** Gated the raw `grailsWebRequest` bean on `@ConditionalOnMissingBean(value = GrailsWebRequestFilter.class, name = "grailsWebRequestFilter")`. When an application overrides only the registration bean, the raw filter now backs off with it rather than lingering unwrapped for Boot to auto-register on `/*` at `LOWEST_PRECEDENCE`. In the default case the raw bean stays a top-level `RequestContextFilter`, so Boot's own `RequestContextFilter` still backs off exactly as before.

And you're right that the old override test only checked bean identity — I reworked it to assemble the real servlet chain via `ServletContextInitializerBeans` and assert against it: the override case verifies no stray `GrailsWebRequestFilter` lands on the chain, and the default case verifies exactly one does. Both new assertions fail without the gating fix (confirmed by reverting it).

codeconsole committed:
Address review nits: drop dependency comment, simplify filter comment

- Remove the empty configuration closure and comment on the spring-test
  compileOnly dependency
- Condense the grailsWebRequest bean comment to the two essential points
  (RequestContextFilter back-off and the registration-bean-name gate)

codeconsole commented:
Acknowledged on sequencing. Worth noting the approach here registers Grails' MVC beans *ahead of* `WebMvcAutoConfiguration` and lets Boot's `@ConditionalOnMissingBean`-guarded beans defer to them, rather than disabling `WebMvcAutoConfiguration` — so it's designed to coexist with code that relies on it being active, not to break it. Happy to verify this branch against the security project before either merges and order them however you'd like — what sequence do you want?

codeconsole committed:
Tighten the filter-chain helper comment

jdaugherty commented:
I missed the dependency when the other PR was merged.  I don't think we want these values registered and instead of docuemnting we need to disable them.

jdaugherty committed:
Add test to show feature changes with EnableWebMvc

jdaugherty commented:
@codeconsole  I pushed the test showing the material differences.  In summary: 

4 failures — each one a Grails 8 behavior change due to removing the annotation:

1. DELETE form body — json.title expected null, got The Stand ([method:DELETE, title:The Stand, pages:1153]): Boot's OrderedFormContentFilter now parses the DELETE body into params.
2. Static resource — /boot-default-mapping.txt expected 404, got 200: Boot's catch-all resource handler now serves classpath:/public at the root.
3. Welcome page — / expected 404, got 200: Boot's WelcomePageHandlerMapping now serves the static index.html.
4. Locale — expected fr_FR (Accept-Language), got de: grails-i18n's SessionLocaleResolver is active, so ?lang=de works.


I think `#4` is expected and a better behavior, so I'm ok with that.  But 1-3 are definitely behavior differences that are material.  We shouldn't be serving assets that Grails doesn't know about nor should we have other mappings.

codeconsole committed:
Clarify §31.3 form-content note: DELETE is the net-new parsed method

PUT/PATCH form fields were already exposed through Grails' params
(GrailsParameterMap); the filter's observable addition for controller
code is DELETE, and it also makes all three visible via the standard
request.getParameter(...) API. Disabling the filter reverts DELETE to
unparsed while PUT/PATCH continue to work.

codeconsole committed:
Suppress Boot's default static-resource and welcome-page handling

With @EnableWebMvc no longer auto-injected, Spring Boot's
WebMvcAutoConfiguration is active and registers a catch-all (/**)
static-resource handler over classpath:/public (etc.) plus welcome-page
mappings that serve a static index.html at '/'. Both overlap with Grails'
own /static/** handler and UrlMappings-based root handling, and change the
404 semantics of unmapped URLs.

Default spring.web.resources.add-mappings to false (overridable) via a new
EnvironmentPostProcessor so Boot's catch-all handler is not registered, and
remove the welcomePageHandlerMapping / welcomePageNotAcceptableHandlerMapping
beans via GrailsWelcomePageAutoConfiguration (opt-out
grails.web.removeWelcomePageMapping) so Grails' UrlMappings own '/'.

codeconsole committed:
Make the form-content filter the single request-body parser for non-POST

Spring's FormContentFilter (registered by Boot's WebMvcAutoConfiguration) parses
form-encoded PUT/PATCH/DELETE bodies at the servlet layer for every Grails
servlet web application, so GrailsParameterMap no longer needs to parse the body
itself — it reads the request parameter map, as it already does for POST. This
removes a duplicate parser that in practice only ran when no filter was present.

- GrailsParameterMap: drop the manual PUT/PATCH form-body parse and its now-unused
  imports/logger; retain the REQUEST_BODY_PARSED constant for binary compatibility.
- GrailsMockHttpServletRequest: parse form-encoded PUT/PATCH/DELETE bodies into the
  parameter map so unit-test mocks reflect a post-filter request (preserves the
  raw-body -> params ergonomic and adds DELETE for parity with runtime).
- GrailsParameterMapTests: exercise the real FormContentFilter (PUT/PATCH/DELETE)
  rather than the removed fallback; GrailsMockHttpServletRequestSpec covers the
  mock's form-body parsing.

With no in-map fallback, spring.mvc.formcontent.filter.enabled=false now fully
disables non-POST form-parameter parsing (single-switch semantics).

codeconsole committed:
Guarantee the form-content filter for @EnableWebMvc applications

The previous commit made Spring's FormContentFilter the single non-POST body
parser, but the filter is only registered by Boot's WebMvcAutoConfiguration —
which backs off when an application declares @EnableWebMvc, leaving PUT/PATCH/
DELETE form bodies unparsed.

GrailsFormContentFilterAutoConfiguration contributes an OrderedFormContentFilter
so a filter is always present. Ordered after WebMvcAutoConfiguration and gated on
@ConditionalOnMissingBean(FormContentFilter) so Boot's takes precedence in a
default app; it only fills the gap when Boot's is absent (the @EnableWebMvc case).
It honours spring.mvc.formcontent.filter.enabled=false so the explicit opt-out
stays a real off-switch. Result: PUT/PATCH/DELETE form parsing is uniform in every
Grails servlet app unless explicitly disabled.

Also:
- enable-mvc-check: the Application class imported @EnableWebMvc but never applied
  it, so the app silently ran as a normal Grails app. Apply the annotation, and
  update WebMvcDefaultsFunctionalSpec to expect DELETE parsed (Grails supplies the
  filter even under @EnableWebMvc) alongside PUT/PATCH.
- upgrading80x.adoc: note that Grails supplies the filter for @EnableWebMvc apps
  and that disabling it turns off PUT/PATCH/DELETE form parsing entirely.

codeconsole committed:
Make the i18n locale resolver strategy configurable

Add grails.i18n.localeResolver (session [default], cookie, acceptHeader,
fixed) so applications can choose how the locale is resolved. The session
and cookie resolvers are mutable and keep the ?lang= LocaleChangeInterceptor;
acceptHeader (the locale follows the Accept-Language header) and fixed are
read-only, so the interceptor is not registered for them. The fixed resolver
uses grails.i18n.default.locale, falling back to the JVM default.

The default remains SessionLocaleResolver, so existing ?lang= switching and
the language selector are unchanged.

codeconsole committed:
Let Grails' locale resolver win under @EnableWebMvc and handle read-only resolvers

Spring 7's WebMvcConfigurationSupport contributes its own localeResolver (an
AcceptHeaderLocaleResolver). Under @EnableWebMvc that bean pre-exists, so
I18nAutoConfiguration's @ConditionalOnMissingBean backed off and Grails' session
resolver never registered, silently disabling ?lang= switching — a regression
from Grails 7, where the i18n resolver always took precedence.

GrailsLocaleResolverAutoConfiguration now removes the WebMvcConfigurationSupport-
contributed localeResolver (identified by its factory type) before
I18nAutoConfiguration is evaluated, so the Grails-configured resolver registers
instead. An application-defined localeResolver is left untouched.

ParamsAwareLocaleChangeInterceptor now handles a read-only LocaleResolver
(AcceptHeaderLocaleResolver/FixedLocaleResolver) gracefully: it logs the
unsupported setLocale at debug rather than an error on every ?lang= request. The
interceptor stays registered so urlMappingsHandlerMapping still has a handler
interceptor; it simply has no effect with a read-only resolver.

The enable-mvc-check app opts into grails.i18n.localeResolver=acceptHeader to
demonstrate header-based resolution cleanly.

codeconsole committed:
Merge branch '8.0.x' into 7.0.x-remove-EnableWebMvc

codeconsole commented:
### Form-content parsing: consolidating on `FormContentFilter`

Following up on @jdaugherty's first point (the newly-active `OrderedFormContentFilter`), I've made the filter the **single** mechanism for parsing non-POST form bodies, rather than having it coexist with `GrailsParameterMap`'s own PUT/PATCH parsing.

**What changed**
- `GrailsParameterMap` no longer parses `PUT`/`PATCH` bodies itself. At runtime the request has already passed through `OrderedFormContentFilter` (order `-9500`, ahead of the Grails and Spring Security filters), so its inline `if (requestMap.isEmpty()) …` parse never actually ran — it only fired in unit tests and filter-absent configs. It's removed; `GrailsMockHttpServletRequest` now simulates the filter so the unit-test ergonomic (set a form body, read `params`) is preserved.
- New `GrailsFormContentFilterAutoConfiguration` guarantees a `FormContentFilter` for every Grails servlet app. Boot registers one for a default app; this only **fills the gap** when Boot's `WebMvcAutoConfiguration` backs off (i.e. an `@EnableWebMvc` app). It's ordered after `WebMvcAutoConfiguration` and gated on `@ConditionalOnMissingBean(FormContentFilter)`, so Boot's takes precedence by default and Grails only supplies one when it's absent. `spring.mvc.formcontent.filter.enabled=false` still disables it — that stays a real off-switch.

Net: `PUT`/`PATCH`/`DELETE` form bodies parse uniformly in every Grails servlet app unless explicitly disabled.

**On `DELETE` — and a heads-up on the `enable-mvc-check` test**

The `enable-mvc-check` `Application` class **imported** `@EnableWebMvc` but never actually applied the annotation (the class only had `@CompileStatic`), so the app was silently running as a normal Grails app — which is why its `DELETE` and locale assertions were failing on the branch. I applied the annotation.

With the annotation applied *and* the fallback filter in place, `DELETE` form bodies are now parsed even under `@EnableWebMvc`, so I flipped the `DELETE` spec from "not parsed" to "parsed." **This intentionally reverses the Grails-7 parity that test asserted, and I want to flag it explicitly.**

The reasoning: parsing a `DELETE` form body is benign and consistent — Boot does it by default, `grails-spring-security` has registered a `FormContentFilter` (which handles DELETE) for years, and the old "`PUT`/`PATCH` yes, `DELETE` no" split was a historical accident in `GrailsParameterMap`, not a deliberate rule. The only argument for *not* parsing it is byte-for-byte Grails-7 parity for the `@EnableWebMvc` escape hatch — i.e. preserving a silently-ignored `DELETE` body that essentially no application relies on. The deliberate change here is specifically that form-content parsing (`PUT`/`PATCH`/`DELETE`) is always on — even under `@EnableWebMvc` — unless disabled via `spring.mvc.formcontent.filter.enabled=false`.

If the consensus is that `@EnableWebMvc` should be exactly Grails 7 on this point too, I'm happy to switch the fallback to a `FormContentFilter` subclass that skips `DELETE` (via `shouldNotFilter`) — but I think uniform parsing is the better default.

**Verification**: `enable-mvc-check` functional spec (7/7), new `GrailsFormContentFilterAutoConfigurationSpec` (5/5), plus `grails-controllers` / `grails-web-common` / `grails-test-suite-uber` / `grails-test-suite-web` green; Checkstyle / CodeNarc / PMD / SpotBugs clean.

codeconsole commented:
## Follow-on changes: neutralizing the Boot MVC defaults exposed by removing `@EnableWebMvc`

Removing the auto-injected `@EnableWebMvc` means Spring Boot's `WebMvcAutoConfiguration` is now active (and Spring's `WebMvcConfigurationSupport` no longer owns MVC config). That's the goal — Grails participates in the Boot auto-configuration ecosystem — but it also switches on a handful of Boot / Spring MVC defaults that overlap with, or silently change, long-standing Grails behavior. These commits keep the pieces Grails should own while letting the rest of `WebMvcAutoConfiguration` through.

### 1. Static resources — don't serve `classpath:/public` at the context root

`WebMvcAutoConfiguration` registers a catch‑all `/**` resource handler over `classpath:/META-INF/resources/`, `classpath:/resources/`, `classpath:/static/` and `classpath:/public/`. Grails already serves those exact locations under `/static/**` (`ControllersAutoConfiguration.GrailsWebMvcConfigurer`), so Boot's `/**` is both redundant and a behavior change: any unmapped URL can now be shadowed by a file under `classpath:/public` where it previously 404'd.

- **`GrailsWebResourcesEnvironmentPostProcessor`** contributes `spring.web.resources.add-mappings=false` as an overridable default, so Boot's catch‑all is off while Grails' own `/static/**` serving (gated by `grails.resources.enabled`) is unaffected. Apps that want Boot's behavior set `spring.web.resources.add-mappings=true`.

### 2. Welcome page — Grails' `UrlMappings` own `/`

`WebMvcAutoConfiguration` also registers `welcomePageHandlerMapping` / `welcomePageNotAcceptableHandlerMapping`, which serve a static `index.html` at `/`. That overlaps with Grails' `UrlMappings`-based root handling and can shadow an application's own `"/"` mapping whenever such an `index.html` happens to be on the classpath.

- **`GrailsWelcomePageAutoConfiguration`** removes both welcome-page mappings so the root path falls through to `UrlMappings`. Opt out with `grails.web.removeWelcomePageMapping=false`.

### 3. Locale resolution — `?lang=` regressed silently under `@EnableWebMvc`

The most subtle one. In **Spring Framework 7, `WebMvcConfigurationSupport` now contributes its own `localeResolver` bean** (an `AcceptHeaderLocaleResolver`). Because `@EnableWebMvc` imports `WebMvcConfigurationSupport`, that bean exists *before* `I18nAutoConfiguration` is evaluated, so grails-i18n's `@ConditionalOnMissingBean(name = "localeResolver")` backed off — Grails' `SessionLocaleResolver` never registered and **`?lang=` locale switching silently stopped working**. Worse, `ParamsAwareLocaleChangeInterceptor` still ran and logged an `UnsupportedOperationException` from `AcceptHeaderLocaleResolver.setLocale` on every `?lang=` request.

This is a regression from Grails 7: on 7.2.x `I18nAutoConfiguration.localeResolver()` registered the `SessionLocaleResolver` **unconditionally**, so it always took precedence even with the auto-injected `@EnableWebMvc`. (Confirmed empirically: dropping the `@ConditionalOnMissingBean` restores `?lang=de` → `de`; keeping it yields the `Accept-Language` header value.)

Fixes:

- **`GrailsLocaleResolverAutoConfiguration`** (ordered before `I18nAutoConfiguration`) removes *only* the `WebMvcConfigurationSupport`-contributed `localeResolver`, identified by its factory-bean type, so grails-i18n's configured resolver registers instead. An application-defined `localeResolver` (from `resources.groovy` or a user `@Configuration`) is left untouched, so an explicit app resolver still wins.
- **`grails.i18n.localeResolver` is now configurable** — `session` (default), `cookie`, `acceptHeader`, or `fixed`. Header-based resolution is a clean opt‑in for apps that prefer it; `acceptHeader`/`fixed` are read-only, so `?lang=` simply has no effect with them.
- **`ParamsAwareLocaleChangeInterceptor`** now handles a read-only resolver gracefully — it logs the unsupported `setLocale` at debug rather than erroring on every request. The interceptor stays registered (so `urlMappingsHandlerMapping` keeps its required `HandlerInterceptor`); it just no-ops when the resolver can't change.

### Tests & docs

- Unit coverage: `GrailsWelcomePageAutoConfigurationSpec`, `GrailsWebResourcesEnvironmentPostProcessorSpec`, `GrailsLocaleResolverAutoConfigurationSpec`, updated `I18nAutoConfigurationSpec`, and a read-only-resolver case in `ParamsAwareLocaleChangeInterceptorTests`.
- The `enable-mvc-check` integration app exercises all of the above end-to-end (`classpath:/public` 404, welcome-page 404, form-content parsing, and — via `grails.i18n.localeResolver=acceptHeader` — header-based locale resolution with no error spam).
- `changingLocales.adoc` documents the new `grails.i18n.localeResolver` property.

jdaugherty commented on grails-controllers/src/main/groovy/org/grails/plugins/web/controllers/GrailsFormContentFilterAutoConfiguration.java:
```diff
@@ -0,0 +1,56 @@
+
+/**
+ * Guarantees a {@link FormContentFilter} for every Grails servlet web application, so form-encoded
+ * {@code PUT}, {@code PATCH} and {@code DELETE} bodies are parsed into request parameters and are
```
The part I did not understand was this is supposed to only apply for `application/x-www-form-urlencoded` request bodies, so I am ok with the changes.  But we should update the documentation so this is clear.

codeconsole replied:
Done in 5c4974389b — the auto-config javadoc and the upgrade notes (section 31.4) now state explicitly that the filter only affects `application/x-www-form-urlencoded` bodies; JSON and multipart bodies are untouched and remain available to data binding.

jdaugherty commented on grails-core/src/main/groovy/grails/config/Settings.groovy:
```diff
@@ -270,6 +270,13 @@ interface Settings {
+     * own the root path ('/') rather than a static {@code index.html} being served for it.
+     * Defaults to true
+     */
+    String WEB_REMOVE_WELCOME_PAGE_MAPPING = 'grails.web.removeWelcomePageMapping'
```
the spring-config-metadata json file needs updated for this new setting & any others that were added.

codeconsole replied:
Done in 5c4974389b — added `grails.web.removeDefaultViewResolverBean`, `grails.web.removeWelcomePageMapping`, and `grails.i18n.localeResolver` (with value hints) to `additional-spring-configuration-metadata.json`.

jdaugherty commented on grails-doc/src/en/guide/upgrading/upgrading80x.adoc:
```diff
@@ -1396,6 +1396,31 @@ Grails 8 registers its request-binding filter as a `RequestContextFilter` bean s
+
+* **`spring.mvc.*` and `spring.web.*` properties now apply.** Properties such as `spring.web.locale`, `spring.mvc.format.date` / `time` / `date-time`, and `spring.web.resources.*` were inert before and now take effect. Review any of these you may have set inadvertently (for example copied from Boot documentation).
+
+* **Boot static-resource handling and welcome page.** Boot adds a catch-all resource handler (`classpath:/META-INF/resources/`, `/resources/`, `/static/`, `/public/`) and a `WelcomePageHandlerMapping` for a static `index.html`, alongside Grails' own resource handling. A request that previously fell through to Grails' URL-mapping error handling may now be served as a static resource or welcome pag ...
```
This file is out of date.

codeconsole replied:
Updated in 5c4974389b. Static resources and the welcome page are now their own subsection (31.3) documenting that Grails removes them by default — they were incorrectly listed as now active. The remaining Boot MVC features moved to 31.4, and a new 31.5 covers locale resolution. Also fixed the stale "two behavioral differences".

jdaugherty commented on grails-doc/src/en/guide/i18n/changingLocales.adoc:
```diff
@@ -28,7 +28,19 @@ Grails will automatically switch the user's locale and subsequent requests will
+        localeResolver: acceptHeader # session (default), cookie, acceptHeader or fixed
+----
+
+`session` and `cookie` are mutable, so the `?lang=` switch works with them. `acceptHeader` (the locale follows the incoming `Accept-Language` header) and `fixed` are read-only, so the `?lang=` switch has no effect and is silently ignored. The `fixed` resolver uses `grails.i18n.default.locale`, falling back to the JVM default locale.
```
So the url will override the header now with your changes? or vice versa?

codeconsole replied:
With the default `session` resolver (and `cookie`), the `?lang=` URL parameter overrides the `Accept-Language` header and is remembered for subsequent requests. With `acceptHeader` and `fixed` the resolver is read-only, so the header (or the fixed locale) wins and `?lang=` is ignored. Spelled this out in the doc in 5c4974389b.

jdaugherty commented on grails-test-examples/enable-mvc-check/README.md:
```diff
@@ -0,0 +1,62 @@
+
+# enable-mvc-check
+
+A dedicated Grails functional test application that pins down the behavior of
```
We should probably remove this file now. It was just here to help you understand the test.

codeconsole replied:
Removed in 5c4974389b.

jdaugherty commented on grails-web-common/src/test/groovy/grails/web/servlet/mvc/GrailsParameterMapTests.groovy:
```diff
@@ -92,82 +97,47 @@ class GrailsParameterMapTests {
-        request.method = 'PUT'
-        request.contentType = "application/x-www-form-urlencoded; charset=UTF-8"
-        request.content = 'foo='.bytes
-        request.removeAttribute(GrailsParameterMap.REQUEST_BODY_PARSED)
```
Since you're relying on the form filter now, you should add a functioanl test to cover these removed scenarios

codeconsole replied:
Added in 5c4974389b — a `FormContentFunctionalSpec` in `app1` (a default, non-@EnableWebMvc app) asserts that form-encoded `PUT`, `PATCH` and `DELETE` (and `POST`) bodies are parsed into `params` via Boot OrderedFormContentFilter, covering these removed unit scenarios end-to-end.

jdaugherty commented on grails-web-common/src/test/groovy/grails/web/servlet/mvc/GrailsParameterMapTests.groovy:
```diff
@@ -92,82 +97,47 @@ class GrailsParameterMapTests {
-        params = new GrailsParameterMap(request)
-        assert params.foo == null // should be null, request can't be parsed twice
-
-        request = new MockHttpServletRequest()
```
Since you're relying on the form filter now, you should add a functioanl test to cover these removed scenarios

codeconsole replied:
Same as above — covered by the new `FormContentFunctionalSpec` in `app1` (5c4974389b).

codeconsole committed:
Retire grails-spring-security's own FormContentFilter; warn when parsing is disabled

Now that Grails guarantees a servlet-level FormContentFilter for every servlet
web application (Boot's OrderedFormContentFilter by default, the Grails fallback
under @EnableWebMvc), it runs at order -9500 — ahead of the whole Spring Security
chain (-100) — and wraps the request before the chain starts. So the plugin's own
in-chain FormContentFilter (inserted at EXCEPTION_TRANSLATION_FILTER.order - 10)
is redundant: the security filters already see the parsed parameters. It was only
needed historically, when Grails 7 auto-injected @EnableWebMvc and no servlet-level
form filter existed.

- SpringSecurityCoreGrailsPlugin: drop the formContentFilter bean registration and
  its import (also removes a bean that Boot could otherwise auto-register on /*).
- SpringSecurityUtils: drop the chain insertion.
- SpringSecurityUtilsIntegrationSpec: update the chain assertions accordingly.

The one config where nothing would then parse non-POST form bodies is an explicit
spring.mvc.formcontent.filter.enabled=false. Rather than have the plugin silently
re-add a filter there (ignoring the opt-out), GrailsFormContentFilterAutoConfiguration
now logs a startup warning when the filter is disabled and none is present, so the
consequence (PUT/PATCH/DELETE form bodies not parsed) is visible.

codeconsole committed:
Address review: docs, config metadata, and functional coverage for form parsing

Responds to review feedback on the @EnableWebMvc removal:

- Docs (upgrading80x.adoc): the upgrade notes were out of date. Boot's static-
  resource handler and welcome page are now removed by default (new section 31.3),
  not "active"; renumber the remaining Boot-MVC notes to 31.4; add 31.5 for locale
  resolution. Clarify that the form-content filter only affects
  application/x-www-form-urlencoded bodies (also in the auto-config javadoc).
- Docs (changingLocales.adoc): spell out precedence — with session/cookie the
  Accept-Language header sets the initial locale and ?lang= overrides and persists
  it; with acceptHeader/fixed the resolver is read-only and ?lang= is ignored.
- Config metadata: add grails.web.removeDefaultViewResolverBean,
  grails.web.removeWelcomePageMapping and grails.i18n.localeResolver (with value
  hints) to additional-spring-configuration-metadata.json.
- Tests: since GrailsParameterMap no longer parses PUT/PATCH bodies itself, add a
  functional test in app1 (a default, non-@EnableWebMvc app) asserting form-encoded
  PUT/PATCH/DELETE/POST bodies are parsed into params via Boot's OrderedFormContentFilter.
- Remove the enable-mvc-check README (it only existed to explain the test).

codeconsole committed:
Fix checkstyle OperatorWrap: put string-concatenation + at end of line

The disabled-filter warning message split its concatenation with a leading +,
which trips Checkstyle's OperatorWrap rule (:grails-controllers:checkstyleMain).

jdaugherty commented on grails-test-examples/spring-security/core/integration-test-app/src/integration-test/groovy/grails/plugin/springsecurity/SpringSecurityUtilsIntegrationSpec.groovy:
```diff
@@ -89,15 +87,14 @@ class SpringSecurityUtilsIntegrationSpec extends AbstractIntegrationSpec {
 
         expect:
-        10 == map.size()
+        9 == map.size()
```
So are we sure the form filter being registered by Grails is ordered before the security interceptor filter?

codeconsole replied:
Yes — and now earlier than before. The plugin used to register `FormContentFilter` inside the chain at `EXCEPTION_TRANSLATION_FILTER.order - 10` (just before `FilterSecurityInterceptor`). The servlet-level `FormContentFilter` runs at order `-9500`, ahead of the whole `springSecurityFilterChain` (`-100`), so it parses the body into params before *any* security filter — including `FilterSecurityInterceptor`. Confirmed by `TestFormParamsControllerSpec` (the putWithParams config): 12/12 with the plugin filter removed, covering `x-www-form-urlencoded` `PUT`/`PATCH` params on a permitAll endpoint (params extracted) and on a secured endpoint (still redirected to login).

jdaugherty commented on grails-controllers/src/main/groovy/org/grails/plugins/web/controllers/GrailsFormContentFilterAutoConfiguration.java:
```diff
@@ -0,0 +1,87 @@
+    @Bean
+    @ConditionalOnMissingBean(FormContentFilter.class)
+    @ConditionalOnBooleanProperty(name = "spring.mvc.formcontent.filter.enabled", matchIfMissing = true)
+    public OrderedFormContentFilter formContentFilter() {
```
Spring Security ran this before the interceptor filter, is this still the case now?

codeconsole replied:
Yes — and now earlier than before. The plugin used to register `FormContentFilter` inside the chain at `EXCEPTION_TRANSLATION_FILTER.order - 10` (just before `FilterSecurityInterceptor`). The servlet-level `FormContentFilter` runs at order `-9500`, ahead of the whole `springSecurityFilterChain` (`-100`), so it parses the body into params before *any* security filter — including `FilterSecurityInterceptor`. Confirmed by `TestFormParamsControllerSpec` (the putWithParams config): 12/12 with the plugin filter removed, covering `x-www-form-urlencoded` `PUT`/`PATCH` params on a permitAll endpoint (params extracted) and on a secured endpoint (still redirected to login).

jdaugherty reviewed (commented):
My only remaining concern is the spring security impacts since it's not clear if the form filter is now applying before or after the spring security filters.
